### PR TITLE
nixos: write our hostname to /etc/hosts

### PR DESCRIPTION
**Pull request by peterhoeg**
###### Motivation for this change

The fqdn of the machine should always be resolvable.
1. set `networking.hostName = "f.q.d.n";`
2. move your machine to a LAN where that machine name is not known
3. run `hostname --fqdn`
4. you will get: `hostname: Name or service not known`

I addition to the `hostname` command, other software that wants to look up the name will also fail without with something like:

```
in `getaddrinfo': getaddrinfo: System error (SocketError)
```

This PR writes out the hostName to /etc/hosts so it is always resolvable.

Additionally, we need to keep it on 2 lines, as `hostname --fqdn` will otherwise return `localhost` instead of the actual name.

I'm not enough of a network guru to see if there are any downsides to this...
###### Things done
- [X] Tested using sandboxing
  ([nix.useChroot](http://nixos.org/nixos/manual/options.html#opt-nix.useChroot) on NixOS,
    or option `build-use-chroot` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
  - [X] NixOS
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by edolstra** [comment]
It's the job of `nss_myhostname` (https://www.freedesktop.org/software/systemd/man/nss-myhostname.html) to ensure that the local hostname always resolves. Unlike a static `/etc/hosts` entry, it will return an IP address on the local network if available and fall back to 127.0.0.1 or ::1 otherwise.

**Comment by peterhoeg** [comment]
I have this in ```/etc/nsswitch.conf``` (excluding non-relevant parts):

```
hosts:     files dns myhostname mymachines
networks:  files dns
```

So maybe it's because the ```dns``` returns an error before ```myhostname``` gets a chance to do its magic?

**Comment by danbst** [comment]
@peterhoeg we've done similar discovery in https://github.com/NixOS/nixpkgs/issues/16754#issuecomment-231713384

**Comment by peterhoeg** [comment]
@edolstra , ```myhostname``` is definitely not doing its job when the hostname is a fully qualified name.

Would you be open to working around it by adding the hostname to ```/etc/hosts```?

**Comment by edolstra** [comment]
I'm not sure if `networking.hostName` is intended to be a fully qualified name. Isn't that what `networking.domain` is for?

In any case, the workaround would completely break nss_myhostname for everybody (since `/etc/hosts` takes precedence), so that's not a good idea.

**Comment by Mic92** [comment]
I think I found a bug: glibc seems not read /etc/nsswitch.conf at all:

```
strace getent hosts hostname 2>&1 | grep nsswitch
```

The output should actually be more like this:

```
$ strace getent hosts <hostname> 2>&1 | grep nsswitch                                                                                  
open("/etc/nsswitch.conf", O_RDONLY|O_CLOEXEC) = 3
read(3, "# Begin /etc/nsswitch.conf\n\npass"..., 512) = 234
```

**Comment by dezgeg** [comment]
At least in the NSCD-disabled case the relevant NSS modules definitely don't get loaded, as glibc doesn't know where to look:
````
[root@jetson-3:~]# strace -f getent hosts $(hostname) 2>&1 | grep nss_
open("/nix/store/ncm44m05zfvj5656xgngc4im1x25iwif-glibc-2.24/lib/libnss_files.so.2", O_RDONLY|O_CLOEXEC) = 3
open("/nix/store/ncm44m05zfvj5656xgngc4im1x25iwif-glibc-2.24/lib/libnss_dns.so.2", O_RDONLY|O_CLOEXEC) = 3
open("/nix/store/ncm44m05zfvj5656xgngc4im1x25iwif-glibc-2.24/lib/libnss_myhostname.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
open("/nix/store/ncm44m05zfvj5656xgngc4im1x25iwif-glibc-2.24/lib/libnss_mymachines.so.2", O_RDONLY|O_CLOEXEC) = -1 ENOENT (No such file or directory)
````

I wonder if we should patch glibc to search for NSS modules from an impure location (e.g. `/run/nss-modules`) managed by a NixOS service (just like we do OpenGL drivers). As I understand it, that would be required anyway to support SSSD as many places on the net recommend not combining SSSD with NSCD.

**Comment by Mic92** [comment]
Yes. I can confirm that glibc cannot find the required nss module

File: nixos/modules/config/nsswitch.nix [hunk 1]
```diff
@@ -10,6 +10,22 @@ let
   inherit (config.services.samba) nsswins;
   ldap = (config.users.ldap.enable && config.users.ldap.nsswitch);
 
+  hostArray = with lib; [
```
Comment (Mic92) [inline nixos/modules/config/nsswitch.nix]: I would `with lib;` to the top of the file.

**Comment by Mic92** [comment]
myhostname and mymachines should work again because https://github.com/NixOS/nixpkgs/issues/21175 was merged.

**Comment by Mic92** [comment]
Thanks. merged - but fixed order of mdns modules.

**Pull request closed without merging**